lib/code_corps/github.ex

```diff
@@ -1,41 +1,245 @@
 defmodule CodeCorps.GitHub do
-  alias CodeCorps.{User, Repo}
 
   @api Application.get_env(:code_corps, :github_api)
 
-  @doc """
-  POSTs `code` to GitHub to receive an OAuth token, then associates the user
-  with that OAuth token.
+  @client_id Application.get_env(:code_corps, :github_app_client_id)
+  @client_secret Application.get_env(:code_corps, :github_app_client_secret)
 
-  Accepts a third parameter – a custom API module – for the purposes of
-  explicit dependency injection during testing.
+  @base_access_token_params %{
+    client_id: @client_id,
+    client_secret: @client_secret
+  }
 
-  Returns one of the following:
+  defmodule APIErrorObject do
+    @moduledoc """
+    Represents an error object from the API.
 
-  - `{:ok, %CodeCorps.User{}}`
-  - `{:error, %Ecto.Changeset{}}`
-  - `{:error, "some_github_error"}`
-  """
-  @spec connect(User.t, String.t, module) :: {:ok, User.t} | {:error, String.t}
-  def connect(%User{} = user, code, api \\ @api) do
-    case code |> api.connect do
-      {:ok, github_auth_token} -> user |> associate(%{github_auth_token: github_auth_token})
-      {:error, error} -> {:error, error}
+    See the type documentation for more about how this struct can be
+    used.
+
+    The full details of error objects can be found in the
+    [GitHub API documentation](https://developer.github.com/v3/#client-errors).
+    """
+    defstruct [:code, :field, :resource]
+
+    def new(opts) do
+      struct(__MODULE__, opts)
+    end
+  end
+
+  defmodule APIError do
+    defstruct [:documentation_url, :errors, :message, :status_code]
+
+    @type t :: %__MODULE__{
+      documentation_url: String.t | nil,
+      errors: List.t | nil,
+      message: String.t | nil,
+      status_code: pos_integer | nil
+    }
+
+    @spec new({integer, map}) :: t
+    def new({status_code, %{"message" => message, "errors" => errors}}) do
+      errors = Enum.into(errors, [], fn error -> convert_error(error) end)
+
+      %__MODULE__{
+        errors: errors,
+        message: message,
+        status_code: status_code
+      }
+    end
+
+    @spec new({integer, map}) :: t
+    def new({status_code, %{"message" => message, "documentation_url" => documentation_url}}) do
+      %__MODULE__{
+        documentation_url: documentation_url,
+        message: message,
+        status_code: status_code
+      }
+    end
+
+    @spec new({integer, map}) :: t
+    def new({status_code, %{"message" => message}}) do
+      %__MODULE__{
+        message: message,
+        status_code: status_code
+      }
+    end
+
+    defp convert_error(%{"code" => code, "field" => field, "resource" => resource}) do
+      APIErrorObject.new([code: code, field: field, resource: resource])
+    end
+  end
+
+  defmodule HTTPClientFailed do
+    defstruct [:reason, message: """
+    The GitHub HTTP client encountered an error while communicating with
+    the GitHub API.
+    """]
```

Comment:
I would love if this would work, but it actually means the string will contain an `\n` after `"with"`. I've been looking for a way to nicely split a string across multiple lines without actually introducing newline characters, and best I could figure out was:

```Elixir
"The GitHub HTTP client encountered an error while communicating with " <>
"the GitHub API."
```

Note the required space after the "with".